internal/command/console_state.go

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) The OpenTofu Authors
+// SPDX-License-Identifier: MPL-2.0
+// Copyright (c) 2023 HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package command
+
+import (
+	"strings"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+)
+
+type consoleBracketState struct {
+	oBrace       int
+	cBrace       int
+	oBracket     int
+	cBracket     int
+	oParentheses int
+	cParentheses int
+	buffer       []string
+}
+
+// BracketsOpen return an int to inform if brackets are open
+// in the console and we should hold off on processing the commands
+// it returns 3 states:
+// -1 is returned the is an incorrect amount of brackets.
+// for example "())" has too many close brackets
+// 0 is returned if the brackets are closed.
+// for examples "()" or "" would be in a close bracket state
+// >=1 is returned for the amount of open brackets.
+// for example "({" would return 2. "({}" would return 1
+func (c *consoleBracketState) BracketsOpen() int {
+	switch {
+	case c.oBrace < c.cBrace:
+		fallthrough
+	case c.oBracket < c.cBracket:
+		fallthrough
+	case c.oParentheses < c.cParentheses:
+		return -1
+	}
+
+	// we calculate open brackets, braces and parentheses by the diff between each count
+	var total int
+	total += c.oBrace - c.cBrace
+	total += c.oBracket - c.cBracket
+	total += c.oParentheses - c.cParentheses
+	return total
+}
+
+// UpdateState updates the state of the console with the latest line data
+func (c *consoleBracketState) UpdateState(line string) {
```

Comment:
What do you think about encapsulating a lot of the bracket and `\\` functionality in here.  The consumers of this only care about: Is there a command I need to execute, What's my indentation level for the prompt?

```
// Returns command if there is one to execute and the level of indentation needed for the prompt
func (c *consoleBracketState) UpdateState(line string) (string, int) {
```

We could then make BracketsOpen and ClearState internal to the package and not used by consumers.  I'd ideally like to encapsulate as much of this as possible.

Reply:
yeah, i can move bracketsOpen in. However, do we not need ClearState external so the consumer makes the decision when to execute and subsequently when to reset?

Reply:
i went with the approach: return the current full command and and open brackets each time we call update. then within update we clear the buffer if open brackets is <= 0